### PR TITLE
Feature/custom decoration painter

### PR DESCRIPTION
**Pull request by flo80**
An example of that I meant in [the discussion](https://github.com/ookami-kb/flutter_arc_text/discussions/14): a simple custom 'decoration painting' method to allow customization of the look of arc text

### CODE REVIEW

File: packages/flutter_arc_text/lib/src/arc_text_painter.dart [hunk 1]
```diff
@@ -79,6 +95,13 @@ class ArcTextPainter {
       ..save()
       ..translate(offset.dx, offset.dy)
       ..rotate(_angleWithAlignment);
+
+    customDecoration?.call(
```
Comment (monsieurtanuki) [inline packages/flutter_arc_text/lib/src/arc_text_painter.dart:99]: Now I understand better ;)
But still, there's one case you cannot deal with here: I want to paint something on the text (like a background, but afterwards) (probably with alpha).

File: packages/flutter_arc_text/example/lib/main.dart [hunk 2]
```diff
@@ -9,6 +9,25 @@ void main() => runApp(const MyApp());
 class MyApp extends StatelessWidget {
   const MyApp({Key? key}) : super(key: key);
 
+  CustomDecorationPainter _drawArc(bool clockwise) =>
+      (Canvas canvas, Rect rect, double finalAngle) {
+        final length = 2 * pi - finalAngle;
+        if (length < 0) return;
+
+        final startAngle =
+            clockwise ? finalAngle - pi / 2 : finalAngle + length + pi / 2;
```
Comment (monsieurtanuki) [inline packages/flutter_arc_text/example/lib/main.dart:18]: We shouldn't see that anymore as you suggested: no more tap dancing with angles (outside of the library, of course).

**Review by monsieurtanuki — COMMENTED** [review]
Hi @flo80! Now that I read your code I understand better what you're up to.

Still not convinced though, mainly because:
* it looks like you still have to compute yourself the start angle (as you do in `main.dart`)
* you cannot draw _on_ the text (possible use-case: hidden text like spoiler), you can only draw _before_ drawing the text
* you don't seem to have control on the displayed height of the letters (which is useful for backgrounds, e.g. 10% padding)

File: packages/flutter_arc_text/example/lib/main.dart (hunk 2 above)
Comment (flo80) [reply packages/flutter_arc_text/example/lib/main.dart:18]: Well I am not sure, if that is fully achievable 🤷  

We could of course change the `CustomDecorationPainter` to provide all angle information (and not just the final angle), e.g. something like this

```
typedef CustomDecorationPainter = void Function(
  Canvas canvas,
  Rect rect,
  double decorationStartingAngle,
  double decorationEndAngle,
);
```

File: packages/flutter_arc_text/lib/src/arc_text_painter.dart (hunk 1 above)
Comment (flo80) [reply packages/flutter_arc_text/lib/src/arc_text_painter.dart:99]: That's true but for 80% of use cases (where you just want to paint generally over the area) you can just put in a stack.

We could of course have a second custom decoration painter which runs after the text has been painted but I feel you than can also switch to the [custom painter](https://github.com/ookami-kb/flutter_arc_text/discussions/14#discussioncomment-588459)  (which could probably shown as an example in the documentation). My goal is mainly to simplify the '80%' of decoration use cases: drawing something along the same circle but not overlapping with the text. Most other things (drawing inside the circle, drawing over the whole thing etc.) can easily be solved with stacks or parent widgets.

Comment (monsieurtanuki) [reply packages/flutter_arc_text/lib/src/arc_text_painter.dart:99]: I don't know if it's 80% of the decoration use cases, and I cannot figure out how to use stacks or parent widgets in order to paint a band on top of a hidden arc text.

This is my actual use case:
* I have a text that I want to display on an arc
* Just before I want to paint a blue band as the text background, starting 10° before, ending 10° after, with a padding of 15 around the text
* could you implement that as an example?

Of course the band hiding the text is definitely an advanced case.

File: packages/flutter_arc_text/example/lib/main.dart (hunk 2 above)
Comment (monsieurtanuki) [reply packages/flutter_arc_text/example/lib/main.dart:18]: Actually I think that if you use the canvas rotation you don't really care about the starting angle (already managed by the canvas).

**Comment by ookami-kb** [comment]
Solved in #17

**Pull request closed without merging**